src/main/java/io/cdap/plugin/salesforce/plugin/source/batch/util/BulkConnectionRetryWrapper.java

```diff
@@ -0,0 +1,209 @@
+      return bulkConnection.getBatchInfoList(jobId);
+    }
+    Object batchInfoList = Failsafe.with(retryPolicy)
+        .onFailure(event -> LOG.info("Failed while getting batch info list"))
+        .get(() -> {
+          try {
+            return bulkConnection.getBatchInfoList(jobId);
+          } catch (AsyncApiException e) {
+            throw new SalesforceQueryExecutionException(e.getMessage());
+          }
+        });
+    return (BatchInfoList) batchInfoList;
+  }
+
+  public BatchInfo getBatchInfo(String jobId, String batchId) throws AsyncApiException {
+    if (!retryOnBackendError) {
+      return bulkConnection.getBatchInfo(jobId, batchId);
+    }
+    Object batchInfo = Failsafe.with(retryPolicy)
+        .onFailure(event -> LOG.info("Failed while getting batc status"))
+        .get(() -> {
+          try {
+            return bulkConnection.getBatchInfo(jobId, batchId);
+          } catch (AsyncApiException e) {
+            throw new SalesforceQueryExecutionException(e.getMessage());
+          }
+        });
+    return (BatchInfo) batchInfo;
+  }
+
+  public InputStream getBatchResultStream(String jobId, String batchId) throws AsyncApiException {
+    if (!retryOnBackendError) {
+      return bulkConnection.getBatchResultStream(jobId, batchId);
+    }
+    Object inputStream = Failsafe.with(retryPolicy)
+        .onFailure(event -> LOG.info("Failed while getting batch result stream"))
+        .get(() -> {
+          try {
+            return bulkConnection.getBatchResultStream(jobId, batchId);
+          } catch (AsyncApiException e) {
+            throw new SalesforceQueryExecutionException(e.getMessage());
+          }
+        });
+    return (InputStream) inputStream;
+  }
+
+  public InputStream getQueryResultStream(String jobId, String batchId, String resultId) throws AsyncApiException {
+    if (!retryOnBackendError) {
+      return bulkConnection.getQueryResultStream(jobId, batchId, resultId);
+    }
+    Object inputStream = Failsafe.with(retryPolicy)
+        .onFailure(event -> LOG.info("Failed while getting query result stream"))
+        .get(() -> {
+          try {
+            return bulkConnection.getQueryResultStream(jobId, batchId, resultId);
+          } catch (AsyncApiException e) {
+            throw new SalesforceQueryExecutionException(e.getMessage());
+          }
+        });
+    return (InputStream) inputStream;
+  }
+
+  public BatchInfo createBatchFromStream(String query, JobInfo job) throws AsyncApiException,
+    SalesforceQueryExecutionException, IOException {
+    if (!retryOnBackendError) {
+      return createBatchFromStreamI(query, job);
+    }
+    Object batchInfo = Failsafe.with(retryPolicy)
+        .onFailure(event -> LOG.info("Failed while creating batch from stream"))
+        .get(() -> {
+          try {
+            return createBatchFromStream(query, job);
+          } catch (AsyncApiException e) {
+            throw new SalesforceQueryExecutionException(e.getMessage());
```

Comment:
SalesforceQueryExecutionException will only be thrown if error code is among the four which we were using previously, write a method that will handle AsyncAPIException and throw error only if code is among those and call that from all catches